### DISCUSSION TITLE
SOLVED   Empty Array

### DISCUSSION THREAD

**Discussion by ghost** [supabase/supabase#3780] [Questions]
SOLVED 

Hello i get empty array when i query my data? 

is anyone know how to fix i
![Ekran Resmi 2021-11-04 10 14 22](https://user-images.githubusercontent.com/42077752/140271819-69bd48d1-0edd-4106-8d14-8281f932679e.png)
t
![Ekran Resmi 2021-11-04 10 15 13](https://user-images.githubusercontent.com/42077752/140271910-f06ba891-6bec-4ce8-a348-b298cc1db6ee.png)
?

**Comment by realPrimoh** [comment]
Well, how did you solve this?

**Reply by stoico** [reply]
Try disabling the RLS on the table as a test, and see if you still get the same [] back.

**Reply by Fahari** [reply]
The problem sadly persists if you are fetching from a database that you do not own. 4 May 22

**Reply by M-a-t-i-n** [reply]
Disabling the RLS solved my problem.

**Reply by joaodk3** [reply]
Hey! 

I advise you to disable the RLS just to test if that is the problem. After this, you can enable it and create a specific Policy for your needs, that would be more secure and appropriate.

You can understand more about Postgres rls and policies here: https://www.postgresql.org/docs/current/ddl-rowsecurity.html

**Reply by shatmanyu** [reply]
Disabling RLS worked!!

**Reply by pcnoic** [reply]
I am not sure that disabling RLS is the go-to option for many of the folk here. Make sure you disable it knowing what potential security implication this might have for your app. 

The OP disabling the RLS will lead to the `listings` table being exposed.

Adding a proper RLS policy instead of disabling it might be the safest path here.

**Reply by Santipac** [reply]
It depends on your policy settings, but in my case I forgot to add the “authenticated” role.

```sql
create policy "Allow authorized get access" 
on public.table for select 
to authenticated, anon
using ( true );

**Reply by mahabubx7** [reply]
> Disabling the RLS solved my problem.

Thanks a lot 🫶
I was debugging and almost lost my mind 😆 
You saved the day 🤝

**Reply by f1yingbanana** [reply]
> It depends on your policy settings, but in my case I forgot to add the “authenticated” role.
> 
> ```sql
> create policy "Allow authorized get access" 
> on public.table for select 
> to authenticated, anon
> using ( true );
> ```

Thank you, you saved me from pain.

**Reply by abdulmajin** [reply]
wow that cool , thanks for this it saves me too

**Comment by jordiyapz** [comment]
For future reader who are beginner in Supabase like I was, if your table has Row Level Security (RLS) enabled, add atleast a policy to allow reading for everyone. There is a template for that.

Check out the [guide](https://supabase.com/docs/guides/auth/row-level-security#allow-read-access)

**Reply by meierpatrick** [reply]
@jordiyapz  Geez, and I was like wtf? :D Thank you for the answer!

**Reply by mattskinosix** [reply]
I love you, thank you

**Reply by jordiyapz** [reply]
Glad to hear that! 😁

**Reply by Sanan4li** [reply]
This thing wasted my 4 hours trying to figure out the issue in my code.

**Reply by edof** [reply]
man, you saved my life. trying to figure what's wrong with my code and you just give me solution

**Reply by TheSlavant** [reply]
what can i say, your advice keeps saving lives

**Reply by abdulmajin** [reply]
thank for notation for beginners

**Comment by sanba-anass** [comment]
disabling rls does not work for me is there any thing that cause this?

**Reply by s0h311** [reply]
Please also destructure the `error` and share the error message with us. So:

```JavaScript
const { data, error } = await supabase.from('blabla').select('name')
console.log(error)
```

**Reply by sarious** [reply]
same issue. Empty data, and error is null. RLS enabled with next lines: 
```
AS PERMISSIVE FOR SELECT
TO public
USING (true)
```

**Reply by richardfwashington** [reply]
Exactly the same for me.
Everything is working fine seemingly.
Connection to supabase, select the table, select all columns, no error, RLS is off (or a policy exists I tried both), there is data in the table.
Returns []

**Reply by greeenboi** [reply]
were you guys able to solve this since its the exact same for me
I'm using NextJS14 with javascript and supabase-js as the lib. following the docs to the letter.
I have tried everything. the funniest part is that i am able to add rows but the api returns an empty array when i try to read it.

**Reply by pandas9** [reply]
> were you guys able to solve this since its the exact same for me I'm using NextJS14 with javascript and supabase-js as the lib. following the docs to the letter. I have tried everything. the funniest part is that i am able to add rows but the api returns an empty array when i try to read it.

yeah lol had same issue just hard refresh

**Reply by richardfwashington** [reply]
I didn't solve it per se.  It just started working.  I'm guessing some kind of caching the needed time to flush.  But to be honest I'm unsure.  If anyone has any light to shed on this I'm sure it would be appreciated.

**Reply by salim-jeylani** [reply]
i solved it by calling `unstable_noStore()` inside my page component. Converting my page component from static to dynamic(no caching). i think.

**Reply by avnogy** [reply]
nextjs 13, solved it by deleting .next/cache, rls was enabled at the start and next cached the empty response

**Reply by devmev10** [reply]
> nextjs 13, solved it by deleting .next/cache, rls was enabled at the start and next cached the empty response

This worked!

**Comment by pandas9** [comment]
for anyone ending up here, for nextjs this is the solution

```
export const supabase = createClientComponentClient({
  options: {
    global: {
      fetch: createFetch({
        cache: "no-store",
      }),
    },
  },
});
```

```
'use server'
export const createFetch =
  (options: Pick<RequestInit, "next" | "cache">) =>
  (url: RequestInfo | URL, init?: RequestInit) => {
    return fetch(url, {
      ...init,
      ...options,
    });
  };
```

**Reply by hmtcelik** [reply]
works like a charm !

**Comment by ronnostic** [comment]
OK, I landed here because I had the same issue:  RLS off, no errors, empty array using `createClient` from `@supabase/supabase-js`

`createClientComponentClient()` from `@supabase/auth-helpers-nextjs` did work, but then I quickly discovered that this package is being deprecated in favor of **`@supabase/ssrs`**

[see this Reddit response](https://www.reddit.com/r/Supabase/comments/17hbwqb/comment/k6sr2b2/?utm_source=share&utm_medium=web2x&context=3)

and this [migration guide](https://supabase.com/docs/guides/auth/server-side/migrating-to-ssr-from-auth-helpers)

Here is a  _minimal_ setup I used to get this working with their new server components package, `@supabase/ssrs`:

### Step1:
uninstall `@supabase/auth-helpers-nextjs`

install `@supabase/ssrs`

### Step 2:

**`env.locaL`:**
```env
NEXT_PUBLIC_SUPABASE_URL=<my-url>
NEXT_PUBLIC_SUPABASE_ANON_KEY=<my-anon-key>
```

### Step 3:

**`src/utils/supabase.ts`:**
```js
import { createServerClient, type CookieOptions } from '@supabase/ssr'

export const supabase = (cookies: CookieOptions) => {
    const cookieStore = cookies()

    return createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
        {
            cookies: {
                get(name: string) {
                    return cookieStore.get(name)?.value
                },
            },
        }
    )
}
```
### Step 4:

**`src/app/myroute/page.tsx`:**
```js
import { supabase } from "@/utils/supabase";
import { cookies } from "next/headers";

export default async function Page() {
  const { data } = await supabase(cookies).from("mytable").select();
  return <pre>{JSON.stringify(data, null, 2)}</pre>;
}

}
```

### Step 5:

Navigate to `http://localhost:3000/myroute` and see your results

### Step 6:
Don't forget to turn RLS back on!

**Reply by loud1990** [reply]
THANK YOU!!!

**Comment by Seokhyeon315** [comment]
I have a same issue. I am using Astro framework with latest version 4.0 and trying to get images from Supabase storage but it returns the empty array [] and error is null. When I look at the supabase dashboard, it seems there was request and response on the server, but I don't know why it returns empty array...

**Reply by jaxkashif34** [reply]
same problem with me. do you find any solution?

**Reply by vice7770** [reply]
Same issue, seems that sometimes it fails and other times comes as []. I think supabase auth(doesn't matter if on the server or not) has some problems. For what i notice in my testing, seems to be affected by high latency.

**Comment by vice7770** [comment]
Hello,
Can you please provide me a time schedule today to talk.
Im available all day.

Thank you
